### PR TITLE
Feature/full rewrite

### PR DESCRIPTION
**Pull request by mbeisel**
Full rewrite finished

### CODE REVIEW

File: src/main/java/com/renderLatex/rest/RenderLatexController.java [hunk 1]
```diff
@@ -125,84 +108,109 @@ public RenderLatexController(ObjectMapper objectMapper, RenderService renderServ
...
-        byte[] pdf = Files.readAllBytes(pdfPath);
-        return pdf;
+    public ResponseEntity<byte[]> renderLatexGet(@RequestParam("packages") List<String> packages, @RequestParam("content") String content, @RequestParam("output") String output) throws IOException {
+        try {
+            for(int i = 0; i < packages.size(); i++){
+                packages.set(i, java.net.URLDecoder.decode(packages.get(i), StandardCharsets.UTF_8.name()));
+            }
+            content = java.net.URLDecoder.decode(content, StandardCharsets.UTF_8.name());
+        } catch (UnsupportedEncodingException e) {
+            e.printStackTrace();
```
Comment (buehlefs) [inline src/main/java/com/renderLatex/rest/RenderLatexController.java]: Answer Request with Error Code 400 (or 500)

File: src/main/java/com/renderLatex/rest/RenderLatexController.java [hunk 2]
```diff
@@ -125,84 +108,109 @@ public RenderLatexController(ObjectMapper objectMapper, RenderService renderServ
...
     )
     public ResponseEntity<byte[]> renderLatex(@RequestBody LatexContent latexContent) throws IOException {
         byte[] renderedFile = null;
+        String filepath ="";
         HttpHeaders headers = new HttpHeaders();
         switch (latexContent.getOutput()){
             case "pdf":
-                this.renderService.renderAsPdf(latexContent);
-                Path pdfPath = Paths.get("renderFile.pdf");
+                filepath = this.renderService.renderAsPdf(latexContent);
```
Comment (buehlefs) [inline src/main/java/com/renderLatex/rest/RenderLatexController.java]: Is it possible to reuse `renderLatexAsPdf` here? This seems like a code duplication.

File: src/main/java/com/renderLatex/rest/RenderLatexController.java [hunk 3]
```diff
@@ -125,84 +108,109 @@ public RenderLatexController(ObjectMapper objectMapper, RenderService renderServ
...
+            }
+            content = java.net.URLDecoder.decode(content, StandardCharsets.UTF_8.name());
+        } catch (UnsupportedEncodingException e) {
+            e.printStackTrace();
+        }
+        LatexContent latexContent= new LatexContent(content, packages, output);
+        byte[] renderedFile = null;
+        String filepath ="";
+        HttpHeaders headers = new HttpHeaders();
+        switch (latexContent.getOutput()){
```
Comment (buehlefs) [inline src/main/java/com/renderLatex/rest/RenderLatexController.java:111]: Refactor both switch case statements into the same method.

File: src/main/java/com/renderLatex/utils/Utils.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import java.io.IOException;
+
+public class Utils {
+    private Utils() {}
+    /*
+    Remove the filename from the filepath (always 15 characters in Latex-Renderer)
+    Then delete the directory and all included files
+     */
+    public static void removeDirectory(String filepath) throws IOException {
+        filepath = filepath.substring(0, filepath.length() - 15);
```
Comment (buehlefs) [inline src/main/java/com/renderLatex/utils/Utils.java]:
```suggestion
        filepath = Paths.get(filepath).getParent().toString();
```

File: src/main/java/com/renderLatex/service/RenderService.java [hunk 5]
```diff
@@ -1,90 +1,139 @@
...
 
 @Service
 public class RenderService {
     private String latexDocClass = "\\documentclass[border=0.50001bp,convert={convertexe={imgconvert},outext=.png}]{standalone} \n";
     private final String docStart = "\\begin{document} \n";
     private final String docEnd = " \\end{document} \n";
-    private String packages = "";
-    private String content = "";
+    private final String userDirectory = Paths.get("").toAbsolutePath().toString();
+    private final String tempDirectory = Paths.get("").toAbsolutePath().toString()+"/tmp";
```
Comment (buehlefs) [inline src/main/java/com/renderLatex/service/RenderService.java]:
```suggestion
    private final String tempDirectory = Paths.get("").toAbsolutePath().resolve("tmp").toString();
```
if the two instance variables are Paths, then all other path manipulation in this module would not have to use string concatenation!

File: src/main/java/com/renderLatex/service/RenderService.java [hunk 6]
```diff
@@ -1,90 +1,139 @@
...
+
+
 
     @Autowired
     public RenderService(){
+        File file = new File(userDirectory +"/tmp");
```
Comment (buehlefs) [inline src/main/java/com/renderLatex/service/RenderService.java]:
```suggestion
        File file = new File(tempDirectory);
```

File: src/main/java/com/renderLatex/service/RenderService.java [hunk 7]
```diff
@@ -1,90 +1,139 @@
...
+        try{
+            FileUtils.deleteDirectory(file);
```
Comment (buehlefs) [inline src/main/java/com/renderLatex/service/RenderService.java:40]: Add a comment for the rationale behind recreating the temp folder from scratch

File: src/main/java/com/renderLatex/service/RenderService.java [hunk 8]
```diff
@@ -1,90 +1,139 @@
...
-        this.content = latexContent.getContent();
-        this.packages = "";
-        latexContent.getLatexPackages().forEach(item -> this.packages += (item + " \n "));
-        createTexDoc();
-        renderTex("renderFile.tex");
+//    private string prepareEnvi
+
+    public String renderAsPdf(LatexContent latexContent){
+        final String uuid = UUID.randomUUID().toString();
+        final String currentImagePath = tempDirectory + "/" + uuid;
```
Comment (buehlefs) [inline src/main/java/com/renderLatex/service/RenderService.java]: Do not use string concatenation to build paths if possible.
Also ideally refactor path creation and folder creation into one helper method.

File: src/main/java/com/renderLatex/service/RenderService.java [hunk 9]
```diff
@@ -1,90 +1,139 @@
...
     }
 
-    public void renderTex(String filename){
+    public void renderTex(String filename, String uuid, String currentImagePath ){
         try {
             ProcessBuilder processBuilderRenderTex = new ProcessBuilder();
             if(System.getProperty("os.name").startsWith("Windows")){
-                processBuilderRenderTex.command("cmd.exe", "/c", "pdflatex -halt-on-error -shell-escape " + filename);
+                System.out.println("pdflatex -halt-on-error -shell-escape " +  uuid + "/"+filename);
+                processBuilderRenderTex.directory(new File(currentImagePath));
```
Comment (buehlefs) [inline src/main/java/com/renderLatex/service/RenderService.java]: move before if-else

File: .gitignore [hunk 10]
```diff
@@ -14,7 +14,7 @@ target/
 .sts4-cache
 
 ### IntelliJ IDEA ###
-.idea
+/.idea
```
Comment (buehlefs) [inline .gitignore]: also add `tmp` document build folder to gitignore

**Review by buehlefs — COMMENTED** [review]
Did not compile for me. I did not test new docker container.

File: README.md [hunk 11]
```diff
@@ -2,7 +2,7 @@
 Service for rendering latex content such as algorithms, diagrams, pictures etc. Latex-Render accepts the document body, the packages that are required to render the document and offers pdf, svg and png as output formats.
 
 ### Version:
-The latest version is: **0.1**
+The latest version is: **1.1**
```
Comment (manuwei) [inline README.md]: as a side comment, we usually prefer to use semantic versioning (x.y.z)_

File: README.md [hunk 12]
```diff
@@ -13,39 +13,32 @@ The latest version is: **0.1**
...
-	2. Make sure all required programs are installed.
 	3. 1st time Setup: *mvn package -DskipTests*
 	4. To start Latex-Renderer run: *mvn spring-boot:run*
 
 ### API:
 Runs on Port 8082.
 
-http://localhost:8082/renderLatex  expects (String content, List<String> latexPackages, String output) as JSON
+It is possible to Request the Data with either a POST- or a GET-Request
+* POST: http://localhost:8082/renderLatex  expects (String content, List<String> latexPackages, String output) as JSON
```
Comment (manuwei) [inline README.md]: I think a short example request for each case would be helpful - for example, how to specify multiple packages (for anyone how is not familiar with springboot and what parameters are expected in this case), what params are required

File: README.md (hunk 11 above)
Comment (manuwei) [reply README.md]: the last version actually is 1.1.0 - just add the last digit :)

**Review by manuwei — CHANGES_REQUESTED** [review]
The code LGTM, I only added a few comments regarding the documentation

Running the tests led to  ``java.lang.Exception: No runnable methods`` - could have a look at that? maybe a @Test annotation or something simliar is missing

**Review by manuwei — APPROVED** [review]

**Pull request merged**